### ISSUE TITLE
accessToken vs. idToken

### ISSUE THREAD

**Issue by crspybits** [AzureAD/microsoft-authentication-library-for-objc#683]
`MSALResult` has both accessToken and idToken properties. It seems like the idToken is a JWT Oauth2 access token. My use case for these tokens is server side. On my server, I have been able to use the accessToken in an https://graph.microsoft.com/v1.0/me call, but the idToken doesn't work in that case. Conversely, with the https://login.microsoftonline.com/common/oauth2/v2.0/token endpoint, the idToken works but the accessToken does not. It look like I have to send both of these tokens to my server as I have use for both of these endpoints. For my interest sake, what is the accessToken? It doesn't seem like an Oauth2 token.

**Comment by oldalton** [comment]
From [here](https://docs.microsoft.com/en-us/azure/active-directory/develop/id-tokens#using-the-id_token): ID Tokens should be used to validate that a user is who they claim to be and get additional useful information about them - it shouldn't be used for authorization in place of an access token. The claims it provides can be used for UX inside your application, as keys in a database, and providing access to the client application. 

From [here](https://docs.microsoft.com/en-us/azure/active-directory/develop/access-tokens): Access tokens enable clients to securely call APIs protected by Azure. Microsoft identity platform access tokens are JWTs, Base64 encoded JSON objects signed by Azure. Clients should treat access tokens as opaque strings, as the contents of the token are intended for the resource only.

To summarize, you should be using accessToken when talking to Microsoft Graph. I'm not entirely clear why would you use idToken with https://login.microsoftonline.com/common/oauth2/v2.0/token endpoint. Can you clarify?
Regarding the format, access token will be in the format needed for your API. It might be JWT, JWE, RPS etc. Your client should treat it as an opaque string and not worry about its format.

**Comment by crspybits** [comment]
Hi Olga,
Thanks for your response. Server-side, I'm using an idToken with https://login.microsoftonline.com/common/oauth2/v2.0/token in the use case of creating a refresh token for offline access-- 
https://docs.microsoft.com/en-us/azure/active-directory/develop/v2-oauth2-on-behalf-of-flow#first-case-access-token-request-with-a-shared-secret
In those docs it wasn't clear to me which kind of token I ought to be using, but if I use an accessToken (and not an idToken), I get:
"AADSTS50027: JWT token is invalid or malformed"

Your information helps-- it wasn't clear to me initially that there were two types of tokens. Right now I'm having one main practical issue. In my server side testing, I need to have a non-expired accessToken and a non-expired idToken. I can generate the accessToken from a refresh token which I have saved for purposes of this testing. But, the idToken involves manual steps in my procedure-- I use an iOS MSAL client to generate that.

**Comment by oldalton** [comment]
Sorry for a late response. I'd recommend to use MSAL for testing and get both id_token and access_token from MSALResult.

**Comment by oldalton** [comment]
Closing this issue, please feel free to reopen if you have additional questions. Thanks!

**Comment by Jins17994** [comment]
@oldalton what is the difference between iDtoken and accesstoken ?  in my case I get only idToken access token is empty.
using JWTDecoder I get the details. and the main issue is cache not clearing, previously login data is found when new customer trying to login.

**Comment by oldalton** [comment]
> From [here](https://docs.microsoft.com/en-us/azure/active-directory/develop/id-tokens#using-the-id_token): ID Tokens should be used to validate that a user is who they claim to be and get additional useful information about them - it shouldn't be used for authorization in place of an access token. The claims it provides can be used for UX inside your application, as keys in a database, and providing access to the client application.
> 
> From [here](https://docs.microsoft.com/en-us/azure/active-directory/develop/access-tokens): Access tokens enable clients to securely call APIs protected by Azure. Microsoft identity platform access tokens are JWTs, Base64 encoded JSON objects signed by Azure. Clients should treat access tokens as opaque strings, as the contents of the token are intended for the resource only.
> 
> To summarize, you should be using accessToken when talking to Microsoft Graph. I'm not entirely clear why would you use idToken with https://login.microsoftonline.com/common/oauth2/v2.0/token endpoint. Can you clarify?
> Regarding the format, access token will be in the format needed for your API. It might be JWT, JWE, RPS etc. Your client should treat it as an opaque string and not worry about its format.

I think links above summarize the difference between those two tokens. Let me know if you have further questions after reviewing the documents.

**Comment by Jins17994** [comment]
thanks for the reply,

this is my code, plz check is there any mistake

*let* viewController = *self* /*replace with your main presentation
controller here */

            *let* webViewParameters =
MSALWebviewParameters(authPresentationViewController: viewController)

            *let* signoutParameters =
MSALSignoutParameters(webviewParameters: webViewParameters)

            signoutParameters.signoutFromBrowser = *false*





            //thisAccount is MSAL Account

            *if* *let* accountToRemove = thisAccount {

            applicationContext!.signout(with: accountToRemove,
signoutParameters: signoutParameters, completionBlock: {(success, error)
*in*



                *if* *let* error = error {

                    print("Couldn't sign out account with error: \(error)")

                    *return*

                }

                print("Sign out completed successfully")

            })

            }*else* {

                print("There is no account to signing out!")

            }

code successfully executed. but again trying to login, then shows previous
logged customer details.

**Comment by oldalton** [comment]
Can you change this line to true?

`signoutParameters.signoutFromBrowser = *true*`

**Comment by Jins17994** [comment]
the above replied mail is my mistake.

**Comment by Jins17994** [comment]
signoutParameters.signoutFromBrowser = *true*

when i replace my code with above mentioned code, after sign out shows a
popup. i click on continue on the popup then very soon it automatically
quit and back to the app. again clicking on the sign in still
shows previous logged customer detail.
could you plz live check my code on google meet or like that? That's
more helpful me and save your valuable time

**Comment by oldalton** [comment]
@Jins17994 - this is B2C, right?

**Comment by Jins17994** [comment]
Azure AD B2C

**Comment by Jins17994** [comment]
The attached file is my sample project. could you please fix the bugs

**Comment by oldalton** [comment]
@Jins17994, I don't think I can see the attachment.
However, implementing signout for B2C is a bit trickier than for AAD because you might have custom policy configuration and not all providers will support signout.

**Comment by Jins17994** [comment]
Thanks for the fast reply,
then how can I sign out from all identity providers in MSAL ADB2C?
In my case click on Facebook then without showing credentials entering the
page and shows previously logged user details.
Is there any way to clear the cache of embedded MSAL webview?

**Comment by oldalton** [comment]
@Jins17994, I believe this question looks similar to yours: https://github.com/MicrosoftDocs/azure-docs/issues/34577

**Comment by Jins17994** [comment]
The above mentioned link is similar to my issue. but there is no Solution
found
Then there is no Solution for sign out from all social accounts in
MSAL😭😭😭?

**Comment by oldalton** [comment]
@Jins17994 - looks like that to me.
If you're using WKWebView, you can try clearing out the cookies yourself.

**Comment by Jins17994** [comment]
Yes, I think about using wkwebview. but how is it possible when after
authentication automatically call back to our app.
